### PR TITLE
GODRIVER-2696 Use an options pattern that doesn't require merging structs

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-2696

## Summary

This PR implements the proposed solution described [here](https://github.com/mongodb/mongo-go-driver/pull/1480#discussion_r1421024693) in the POC PR #1480 with the intention of updating the options pattern to abstract merging.

## Background & Motivation

To avoid creating merge function for each option type, the goal of 2696 is to find and implement an optional arguments pattern that doesn't require maintaining dedicated code for merging each type of option struct together when multiple options structs are provided.

### CODE REVIEW

File: internal/integration/unified_spec_test.go [hunk 1]
```diff
@@ -282,35 +284,52 @@ func runSpecTestCase(mt *mtest.T, test *testCase, testFile testFile) {
...
 		})
 		testClientOpts.SetServerMonitor(test.monitor.sdamMonitor)
-		if testClientOpts.HeartbeatInterval == nil {
+		if args.HeartbeatInterval == nil {
 			// If one isn't specified in the test, use a low heartbeat frequency so the Client will quickly recover when
 			// using failpoints that cause SDAM state changes.
 			testClientOpts.SetHeartbeatInterval(defaultHeartbeatInterval)
 		}
+
+		//newOpts := mongoutil.NewOptionsFromArgs[options.ClientArgs](args, nil)
```
Comment (qingyang-hu) [inline internal/integration/unified_spec_test.go:332]:
```suggestion
```

File: mongo/client_encryption.go [hunk 2]
```diff
@@ -94,12 +80,19 @@ func NewClientEncryption(keyVaultClient *Client, opts ...*options.ClientEncrypti
...
+	db *Database, coll string, createOpts Options[options.CreateCollectionArgs],
 	kmsProvider string, masterKey interface{}) (*Collection, bson.M, error) {
 	if createOpts == nil {
 		return nil, nil, errors.New("nil CreateCollectionOptions")
 	}
-	ef := createOpts.EncryptedFields
+
+	createArgs, err := newArgsFromOptions[options.CreateCollectionArgs](createOpts)
+	if err != nil {
+		fmt.Println("err: ", err)
```
Comment (qingyang-hu) [inline mongo/client_encryption.go:91]:
```suggestion
```

File: mongo/util.go
Comment (qingyang-hu) [inline mongo/util.go:1]: We can use `ptrutil.Ptr` instead.

**Review by qingyang-hu — APPROVED** [review]
LGTM!

File: internal/mongoutil/mongoutil.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+		}
+	}
+	return args, nil
+}
+
+// Options implements a mongo.Options object for an arbitrary arguments type.
+// The intended use case is to create options from arguments.
+type Options[T Args] struct {
+	Args *T             // Arguments to set on the option type
+	Clbk func(*T) error // A callback for further modification
```
Comment (matthewdale) [inline internal/mongoutil/mongoutil.go:78]: Optional: It's not obvious that `Clbk` is an abbreviation for "Callback" because it's similar to other more common abbreviations (e.g. `clk`, `blk`). Consider renaming to `Callback`.

File: mongo/options.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+package mongo
+
+import (
+	"go.mongodb.org/mongo-driver/internal/mongoutil"
+)
+
+// Options is an interface that wraps a method to return a list of setter
+// functions that can set a generic arguments type.
+type Options[T mongoutil.Args] interface {
```
Comment (matthewdale) [inline mongo/options.go:15]: Is there a risk to using `T any` instead of `T mongoutil.Args`?

File: mongo/options.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	ArgsSetters() []func(*T) error
```
Comment (matthewdale) [inline mongo/options.go:16]: Is there a reason to return errors from the setters? Handling the errors adds a significant amount of code and is currently completely unused. Do we plan to use the ability to return errors in the future for something?

File: mongo/options/clientoptions.go [hunk 6]
```diff
@@ -258,328 +256,381 @@ type ClientOptions struct {
...
+	return c.Opts
 }
 
-func (c *ClientOptions) validate() error {
-	// Direct connections cannot be made if multiple hosts are specified or an SRV URI is used.
-	if c.Direct != nil && *c.Direct {
-		if len(c.Hosts) > 1 {
+// ValidateClientArgs checks if the client arguments will create a valid
+// connection.
+func ValidateClientArgs(args *ClientArgs) error {
```
Comment (matthewdale) [inline mongo/options/clientoptions.go:285]: It seems like this is separate from `ClientOptions.Validate` so it can also be called in the `topology` package for a `ClientArgs`. Can we move the `Validate` method to a `ClientArgs` instead so we don't have to add this exported symbol?

Alternatively, we should move this validation to an internal package and remove `ClientOptions.Validate`.

File: mongo/options.go (hunk 4 above)
Comment (qingyang-hu) [reply mongo/options.go:15]: Although, functionally, `T any` fits here as well, I think it's better to restrict the types. The `mongoutil.Args` type also implies the expected usage. The only issue is that we need to maintain the `Args` list.

Comment (matthewdale) [reply mongo/options.go:15]: I'm concerned about the cost of maintaining that exhaustive list and also of using something defined in an internal package in the `options` API. The usage of `options.Options` in the `mongo` and `options` APIs seems pretty clear already.

Comment (prestonvasquez) [reply mongo/options.go:15]: No matter what, we would need a compile-time check that options implement `ArgsSetters`. `mongoutil.Args` serves this purpose under the proposed paradigm, as well as providing type safety to the `mongo.Options` type. 

In terms of maintaining the list, options are part of stable API so this list could only grow within the scope of 2.x.

Unfortunately, due to dependency cycles we can't define this list in the mongo package. However, it doesn't really matter: if an embedded type (WLOG consider options.ClientArgs) is removed from `mongoutil.Args`, then you will get a compile-time failure akin to the following:

```
"go.mongodb.org/mongo-driver/mongo/options".ClientArgs does not satisfy mongoutil.Args
```

The _only_ way you wouldn't get this failure is if (1) the underlying options are not used [in which case removal doesn't matter], or (2) the options have been removed in all instance [a breaking change].

IMO the benefit of type-safety outweighs the need to extend this list when new option requirements arise.

Comment (matthewdale) [reply mongo/options.go:15]: I was actually thinking about when we need to add new `Options` types, which would require adding a type to the `mongoutil.Args` interface. For example, if we add a new `Collection` method `coll.Foo`, we'd probably also add `FooOptions` and `FooArgs`, which will need to implement `Options`. We'd need to add `FooArgs` to `mongoutil.Args`. Could that change break any user code?

Another consideration is users who want to write functions that operate on `Options`. For example, if a user wanted to pass `*Args` types directly to collection methods (like we do in our spec tests), they might define a generic type like:
```go
type argsAsOptions[T mongoutil.Args] struct {
	args *T
}

func (o *argsAsOptions[T]) ArgsSetters() []func(*T) error {
	return []func(*T) error{
		func(args *T) error {
			*args = *o.args
			return nil
		},
	}
}

func main() {
	coll.FindOne(
		context.Background(),
		bson.D{},
		&argsAsOptions[options.FindOneArgs]{
			args: &options.FindOneArgs{
				Comment: "blah",
			},
		},
	)
}
```

However, a user can't reference `mongoutil.Args` from outside the Go Driver module, so they would have to define a type constraint that is a copy of `mongoutil.Args`:
```go
type myArgs interface {
	options.AggregateArgs | options.BucketArgs | options.BulkWriteArgs |
	// ...
}

type argsAsOptions[T myArgs] struct {
	// ...
```
That example is inspired by a public code search for users defining `ClientOptions` literals instead of using the builder API (see [here](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+repo:has.content%28go.mongodb.org/mongo-driver%29+-file:%28%5E%7C/%29vendor/+%26options.ClientOptions&patternType=standard&case=yes&sm=0&groupBy=repo)). Is that a use case we want to support?

Concerning type safety, is there a problematic case that using the `mongoutils.Args` type constraint prevents?

Comment (prestonvasquez) [reply mongo/options.go:15]: > I was actually thinking about when we need to add new Options types, which would require adding a type to the mongoutil.Args interface. 

We do this as a compile-time check, it's essentially this:

```go
var _ mongoutil.Args = &FindOneArgs{}
```

Removing any arg struct (that is actually being used somewhere in the code) from `mongoutil.Args` will result in the following error [WLOG]:

```
"go.mongodb.org/mongo-driver/mongo/options".ClientArgs does not satisfy `mongoutil.Args`
```

> Could that change break any user code?

Users don't need to be able to access or copy `mongoutil.Args`. Although it's unclear to me why you would want to do this, if a user is trying to genericize their own options they can create a type that implements a subset of `mongoutil.Args`, e.g.:

```go
type opts interface {
	options.FindOneArgs | options.InsertOneArgs | SomeRandomStruct
}

type argsAsOptions[T opts] struct {
	args *T
}
```

In general, I would expect users to define custom types without generics:

```go
type argsAsOptions struct {
	args *options.FindOneArgs
}

func (o *argsAsOptions) ArgsSetters() []func(*options.FindOneArgs) error {
	return []func(*options.FindOneArgs) error{
		func(args *options.FindOneArgs) error {
			*args = *o.args

			return nil
		},
	}
}

func main() {
	// ...

	coll.FindOne(
		context.Background(),
		bson.D{},
		&argsAsOptions{
			args: &options.FindOneArgs{
				Comment: "blah",
			},
		},
	)
}
```

> That example is inspired by a public code search for users defining ClientOptions literals instead of using the builder API (see [here](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+repo:has.content%28go.mongodb.org/mongo-driver%29+-file:%28%5E%7C/%29vendor/+%26options.ClientOptions&patternType=standard&case=yes&sm=0&groupBy=repo)). Is that a use case we want to support?

I would also expect users to define struct literals by appending the `Opts` field:

```go
opts := options.Client().ApplyURI("mongodb://localhost:27017")

opts.Opts = append(opts.Opts, func(args *options.ClientArgs) error {
	appName := "myApp"

	*args = options.ClientArgs{
		AppName: &appName,
	}

	return nil
})

client, err := mongo.Connect(opts)
if err != nil {
	panic(err)
}
```

> Concerning type safety, is there a problematic case that using the mongoutils.Args type constraint prevents?

Probably not, since we define the expected options type in the function signature, preventing any external arguments being used in our code:

```go
func (coll *Collection) InsertOne(ctx context.Context, document interface{},
	opts ...TOptions[options.InsertOneArgs]) (*InsertOneResult, error) {
```

---

I don't anticipate users ever needing to directly interact with `mongo.Options`. IIUC strongly typing the the arg setter parameters will not cause issues, and will result in the same behavior as `any`. If you feel strongly about using `any` I'm happy to make that change.

Comment (matthewdale) [reply mongo/options.go:15]: Given the examples you provided which don't require generics or copying `mongoutil.Args`, I agree it doesn't block any known use cases. If a user did want to define a generic method for converting an `*Args` into an `Options`, defining a type constraint that is a subset of `mongoutil.Args` is somewhat inconvenient, but also not blocking. Ideally, if we think that's a use case that many users are interested in, we can provide that utility in the `options` package in the future.

Consider this resolved.

File: mongo/options.go (hunk 5 above)
Comment (prestonvasquez) [reply mongo/options.go:16]: The setter error is used on L605 of the client options, in the `ApplyURI` method.

Comment (matthewdale) [reply mongo/options.go:16]: Ah, good point. Considering we're going to be stuck with this API for at least all of v2.x, and there's an example of it actually being used, the added flexibility of being able to use different setter validation patterns in the future seems worth the additional error handling code, even if it remains mostly unused.

Consider this resolved.

File: internal/mongoutil/mongoutil.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+			}
+
+			return nil
+		},
+	}
+}
+
+// NewOptionsFromArgs will construct an Options object from the provided
+// arguments object.
+func NewOptionsFromArgs[T Args](args *T, clbk func(*T) error) *Options[T] {
```
Comment (matthewdale) [inline internal/mongoutil/mongoutil.go:102]: Optional: The `clbk` argument seems unused and can be removed.

File: x/mongo/driver/topology/topology_options.go [hunk 8]
```diff
@@ -41,41 +42,60 @@ type Config struct {
...
-	if s.DeprecationErrors != nil {
-		driverOpts.SetDeprecationErrors(*s.DeprecationErrors)
+	if args.DeprecationErrors != nil {
+		driverOpts.SetDeprecationErrors(*args.DeprecationErrors)
 	}
 	return driverOpts
 }
 
-func newLogger(opts *options.LoggerOptions) (*logger.Logger, error) {
+func newLogger(opts mongoutil.MongoOptions[options.LoggerArgs]) (*logger.Logger, error) {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply x/mongo/driver/topology/topology_options.go:58]: Much of the stable API function signatures used in the internal and x/ packages are defined with `Options`. Converting options to args too quickly can be equally awkward. Consider the use case with `runTest` [here](https://github.com/mongodb/mongo-go-driver/pull/1570/files#diff-279618ef31bd58babcb790025dafc4825d9e6b6a7eed4e2e8b95d9e23717efffR55). Part of the logic for `runTest` is to create a mongo client:

```go 
func runTest(ctx context.Context, clientOpts *options.ClientOptions) error {
	client, err := mongo.Connect(clientOpts)
	if err != nil {
		return fmt.Errorf("Connect error: %w", err)
	}

	// ...
}
```

If we were to update the function signature to use args, then we would have to convert the args back into options:

```go
func runTest(ctx context.Context, clientArgs *options.ClientArgs) error {
	clientOpts := mongoutil.NewOptionsFromArgs(clientArgs, nil)

	client, err := mongo.Connect(clientOpts)
	if err != nil {
		return fmt.Errorf("Connect error: %w", err)
	}
	
	// ...
}
```

The same is true for the integration function`executeAdminCommandWithRetry`, which passes options to `RunCommand`. This is the only other instance of this pattern I could find in the proposal.

File: mongo/options/encryptoptions.go [hunk 9]
```diff
@@ -16,17 +16,35 @@ const (
...
 }
 
-// EncryptOptions represents options to explicitly encrypt a value.
-type EncryptOptions struct {
+// RangeOptions contains options to configure RangeArgs for queryeable
+// encryption. Each option can be set through setter functions. See
+// documentation for each setter function for an explanation of the option.
+type RangeOptions struct {
+	Opts []func(*RangeArgs) error
+}
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/options/encryptoptions.go:34]: If the specifications are ever extended to include an API that accepts `RangeOptions` as a parameter, then we would be forced to create a duplicate type that fits the functional pattern.

File: mongo/gridfs_bucket.go [hunk 10]
```diff
@@ -190,26 +190,23 @@ func (b *GridFSBucket) DownloadToStream(ctx context.Context, fileID interface{},
...
-	var numSkip int32 = -1
-	var sortOrder int32 = 1
+	args, err := newArgsFromOptions[options.GridFSNameArgs](opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct arguments from options: %w", err)
+	}
 
-	nameOpts := options.GridFSName()
-	nameOpts.Revision = &options.DefaultRevision
+	var numSkip int32 = -1
```
Comment (qingyang-hu) [inline mongo/gridfs_bucket.go:200]: Shall we initialize `numSkip` with `&options.DefaultRevision`?

Comment (prestonvasquez) [reply mongo/gridfs_bucket.go]: That makes sense. Looking at this method a bit closer, it's unclear why we set the Revision on the arguments. Suggest we also remove that code.

**Review by qingyang-hu — DISMISSED** [review]

File: x/mongo/driver/topology/topology_options.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply x/mongo/driver/topology/topology_options.go:58]: Sounds good. Consider this comment resolved.

File: mongo/options/encryptoptions.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/options/encryptoptions.go:34]: I think this should be resolved by the proposed change to use the `*Options` and `*OptionsBuilder` names. In that case, we can add a new type `RangeOptionsBuilder` that would be used if there was ever a need to add varargs-style `RangeOptions`.

File: internal/mongoutil/mongoutil.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+			}
+
+			return nil
+		},
+	}
+}
+
+// NewOptionsFromArgs will construct an Options object from the provided
+// arguments object.
+func NewOptionsFromArgs[T Args](args *T) *ArgOptions[T] {
```
Comment (matthewdale) [inline internal/mongoutil/mongoutil.go:102]: Optional: Consider making this return a `mongoutil.Options`, which would allow un-exporting the `ArgOptions` type.

```suggestion
func NewOptionsFromArgs[T Args](args *T) Options[T] {
```

File: internal/mongoutil/mongoutil.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+		options.RewrapManyDataKeyArgs | options.RunCmdArgs |
+		options.SearchIndexesArgs | options.ServerAPIArgs | options.SessionArgs |
+		options.TimeSeriesArgs | options.TransactionArgs | options.UpdateArgs |
+		options.UpdateSearchIndexArgs | options.ChangeStreamArgs |
+		options.AutoEncryptionArgs
+}
+
+// MongoOptions is an interface that wraps a method to return a list of setter
+// functions for merging options for a generic argument type.
+type MongoOptions[T Args] interface {
```
Comment (matthewdale) [inline internal/mongoutil/mongoutil.go:42]: Optional: Consider renaming this to `Options` to remove the redundant naming.

```suggestion
type Options[T Args] interface {
```

**Review by matthewdale — COMMENTED** [review]
Looks good, pending the `*Args` -> `*Options` and `*Options` -> `*OptionsBuilder` naming changes.

File: internal/mongoutil/mongoutil.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+				return nil, err
+			}
+		}
+	}
+	return args, nil
+}
+
+// Options implements a mongo.Options object for an arbitrary arguments type.
+// The intended use case is to create options from arguments.
+type Options[T Args] struct {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go]: I think renaming is a good idea, but note that a rename will not fix all literal cases. For example the user would still need to update accordingly for this case:

```go
loggerOptions := &options.LoggerOptions{}

// compile error: expecting options.LoggerOptionsBuilder
clientOpts.SetLoggerOptions(loggerOptions) 
```

To update:
```go
bldr := &options.LoggerOptionsBuilder{
	Opts: []func(*options.LoggerOptions) error {
		func(lopts *options.LoggerOptions) error {
				loggerOptions := &options.LoggerOptions{}
				*lopts = *loggerOptions

				return nil
			},

	},
}

clientOpts.SetLoggerOptions(bldr)
```

File: internal/mongoutil/mongoutil.go (hunk 11 above)
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go]: Options (now OptionsBuilder) is an interface. I think it would be better to export `OptionsBuilderWithCallback`. We can update this later if required considering it's internal.

File: internal/mongoutil/mongoutil.go (hunk 12 above)
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go]: I updated it to OptionsBuilder, since it interfaces that pattern.

File: mongo/options/encryptoptions.go [hunk 14]
```diff
@@ -16,39 +16,78 @@ const (
...
-// EncryptOptions represents options to explicitly encrypt a value.
+// RangeOptionsBuilder contains options to configure Rangeopts for queryeable
+// encryption. Each option can be set through setter functions. See
+// documentation for each setter function for an explanation of the option.
+type RangeOptionsBuilder struct {
+	Opts []func(*RangeOptions) error
+}
+
+// Range creates a new RangeOptions instance.
+func Range() *RangeOptionsBuilder {
```
Comment (matthewdale) [inline mongo/options/encryptoptions.go:37]: These are unused and should be removed. We can add them in the future if they become necessary.

**Review by matthewdale — CHANGES_REQUESTED** [review]

File: internal/mongoutil/mongoutil.go [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+type OptionsBuilder[T Options] interface {
+	OptionsSetters() []func(*T) error
+}
+
+// NewOptionsFromBuilder will functionally merge a slice of mongo.Options in a
+// "last-one-wins" manner, where nil options are ignored.
+func NewOptionsFromBuilder[T Options](opts ...OptionsBuilder[T]) (*T, error) {
+	args := new(T)
+	for _, opt := range opts {
+		if opt == nil || reflect.ValueOf(opt).IsNil() {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go:52]: Can you provide documentation concerning the heap allocation for structs with use of `reflect.ValueOf`? It's not obvious to me why this function would need to allocate anything to heap in this case. For example:

```
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/internal/mongoutil
BenchmarkReflectValueOf-10       1602472               635.8 ns/op          1428 B/op         11 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/internal/mongoutil  5.912s
```

```go
func BenchmarkReflectValueOf(b *testing.B) {
	s := make([]OptionsBuilder[options.ClientOptions], 0, 1_000_000)
	for i := 0; i < 1_000_000; i++ {

		s = append(s, options.Client().ApplyURI("x").SetAppName("y").
			SetAuth(options.Credential{}).SetHosts([]string{"x", "y"}).
			SetDirect(true).SetTimeout(time.Second))
	}

	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_, _ = NewOptionsFromBuilder[options.ClientOptions](s...)
	}
}
```

Here are the results when you check for `nil` in the options setter:

```
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/internal/mongoutil
BenchmarkReflectValueOf-10       1396346               774.1 ns/op          1579 B/op         17 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/internal/mongoutil  10.938s
```

[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go:52]: Can you provide documentation concerning the heap allocation for structs with use of `reflect.ValueOf`? It's not obvious to me why this function would need to allocate anything to heap:

```go
func BenchmarkReflectValueOf(b *testing.B) {
	s := make([]OptionsBuilder[options.LoggerOptions], 0, 1_000_000)
	for i := 0; i < 1_000_000; i++ {
		var lo *options.LoggerOptionsBuilder

		s = append(s, lo)
	}

	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_, _ = NewOptionsFromBuilder[options.LoggerOptions](s...)
	}
}
```
```
❯ go1.18.1 test -bench=. -benchmem
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/internal/mongoutil
BenchmarkReflectValueOf-10           499           2122966 ns/op           32123 B/op          1 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/internal/mongoutil  1.996s
```

Additionally, checking for nil in `OptionsSetter()` does not actually solve this issue. The `reflect.IsNil()` check is to insure that an interface containing an nil concrete type doesn't call `OptionsSetters()`. For example:

```go
var lo *options.LoggerOptionsBuilder
_, _ = NewOptionsFromBuilder[options.LoggerOptions](s...)
``` 

That is, the nil pointer dereference doesn't happen _in_ `OptionsSetters`, it happens _calling_ `OptionsSetters`. The _in_ case is covered by the first check in the condition `opt == nil`, ensuring `reflect.ValueOf` is not called.

[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go:52]: Also, I wouldn't expect reflect.ValueOf to functionally allocate to stack over heap for structs,  unless they are massive. For example:

```go
func BenchmarkReflectValueOf(b *testing.B) {
	b.ReportAllocs()

	example := options.Client().
		ApplyURI("x").
		SetAppName("y").
		SetAuth(options.Credential{})

	for i := 0; i < b.N; i++ {
		_ = reflect.ValueOf(example).IsNil()
	}
}
```

```
❯ go1.18.1 test -bench=. -benchmem
goos: darwin
goarch: arm64
pkg: heap
BenchmarkReflectValueOf-10      714014944                1.615 ns/op           0 B/op          0 allocs/op
PASS
ok      heap    2.033s
```

[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go:52]: I see the update you're referencing: https://go-review.googlesource.com/c/go/+/408826?usp=email 

I'm not sure determining how to avoid using reflection here is worth the optimization at the moment.

File: mongo/options.go [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+// newOptionsFromBuilder wraps the given mongo-level options in the internal
+// mongoutil options, merging a slice of options in a last-one-wins algorithm.
+func newOptionsFromBuilder[T mongoutil.Options](opts ...Options[T]) (*T, error) {
+	mongoOpts := make([]mongoutil.OptionsBuilder[T], len(opts))
+	for idx, opt := range opts {
+		mongoOpts[idx] = opt
+	}
+
+	return mongoutil.NewOptionsFromBuilder(mongoOpts...)
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/options.go:27]: Go functions should compile inline under performance-critical conditions, see [here](https://dave.cheney.net/2020/04/25/inlining-optimisations-in-go). Benchmarking direct inlining and compiler inlining yields virtually the same results:

Indirect:
```
❯ go1.18.1 test -bench=BenchmarkNewArgsFromOptions -benchmem
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/mongo
BenchmarkNewArgsFromOptions-10          17959299                67.02 ns/op           64 B/op          2 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/mongo       8.933s
```

Direct:
```
❯ go1.18.1 test -bench=BenchmarkNewArgsFromOptions -benchmem
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/mongo
BenchmarkNewArgsFromOptions-10          18750670                67.71 ns/op           64 B/op          2 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/mongo       9.365s
```

Adding `//go:noinline` to `NewOptionsFromBuilder`:
```
❯ go1.18.1 test -bench=BenchmarkNewArgsFromOptions -benchmem
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/mongo
BenchmarkNewArgsFromOptions-10          16543612                71.57 ns/op           64 B/op          2 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/mongo       8.454s
```

File: internal/integration/mtest/setup.go [hunk 17]
```diff
@@ -57,15 +58,20 @@ var testContext struct {
...
+func setupClient(opts *options.ClientOptionsBuilder) (*mongo.Client, error) {
+	args, err := mongoutil.NewOptionsFromBuilder[options.ClientOptions](opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct options from builder: %w", err)
+	}
+
 	wcMajority := writeconcern.Majority()
 	// set ServerAPIOptions to latest version if required
-	if opts.ServerAPIOptions == nil && testContext.requireAPIVersion {
+	if args.ServerAPIOptions == nil && testContext.requireAPIVersion {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/integration/mtest/setup.go:69]: This is not a particularly simple change. The renaming has made variable declaration and function signatures concerning builders and options awkward. For example, 

```go
func (coll *Collection) InsertOne(ctx context.Context, document interface{},
	opts ...Options[options.InsertOneOptions]) (*InsertOneResult, error) {

	args, err := newOptionsFromBuilder(opts...)
	if err != nil {
		return nil, err
	}
	
	// ...
}
```

In this case we retain the function signature `opts` argument, even though technically the representation is an OptionsBuilder. Using `optsBldrs` or `bldrs` in place of this argument looks strange and could be confusing for users. But what should the internal logic of the function, which uses the builder to construct an actual Options struct, call Options? `args` seems to make sense as options is synonymous with a list of arguments. Unfortunately we can't reuse `opts`. Open to suggestions, I think we should keep the naming consistent, however. So in the case you point out `opts` works but we probably shouldn't use the variables interchangeably throughout the code base.

File: mongo/options/encryptoptions.go (hunk 14 above)
Comment (prestonvasquez) [reply mongo/options/encryptoptions.go:37]: RangeOptionsBuilder is used to create explicit options for FLE: https://github.com/mongodb/mongo-go-driver/blob/1b7cde6e7d946cc66da0c9ec57fcffc1a89e2877/mongo/client_encryption.go#L249

File: internal/mongoutil/mongoutil.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+			}
+		}
+	}
+	return args, nil
+}
+
+// OptionsBuilderWithCallback implements a mongo.OptionsBuilder object for an
+// arbitrary options type. The intended use case is to create options from
+// options.
+type OptionsBuilderWithCallback[T Options] struct {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go:75]: There is already a constructor for this, `NewBuilderFromOptions`. I can extend it to include callbacks, but the reason we export `OptionsBuilderWithCallback` is because it makes it easier to create a slice of builders with callbacks: https://github.com/prestonvasquez/mongo-go-driver/blob/1248c26ff39d103f56505d6261fb99e7173cacf4/internal/integration/unified/collection_operation_execution.go#L1179-L1208

I can extend `mongoutil.NewBuilderFromOptions` to include a callback option, however.

[the remark this answers is not in the document]
Comment (prestonvasquez) [reply internal/mongoutil/mongoutil.go:75]: I think we should still export and return the OptionsBuilderWithCallback struct, so as not to break the dependency inversion principle. I will extend the existing `NewBuilderFromOptions` function signature to include a callback argument.

File: mongo/options.go [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"go.mongodb.org/mongo-driver/internal/mongoutil"
+)
+
+// Options is an interface that wraps a method to return a list of setter
+// functions that can set a generic arguments type.
+type Options[T mongoutil.Options] interface {
+	OptionsSetters() []func(*T) error
+}
```
Comment (matthewdale) [inline mongo/options.go:17]: I think it's worth doing a PoC of moving this interface into the `options` package. That may solve a number of problems we're running into with passing options types into other options setters.

File: mongo/options.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/options.go:27]: The optimization is mostly from not having to create the `mongoOpts` copy. However, that optimization is almost unnoticeable when considered in the perspective of an entire operation, so it's probably not worth doing. Here's a quick benchmark of calling `FindOne` with 5 option setters using the `mongoOpts` copy vs copy/pasting the code inline:
```
❯ benchstat call.txt inline.txt                                                
goos: darwin
goarch: arm64
pkg: loadtest
          │  call.txt   │          inline.txt           │
          │   sec/op    │   sec/op     vs base          │
Client-10   87.04µ ± 2%   86.61µ ± 1%  ~ (p=0.579 n=10)

          │   call.txt   │             inline.txt              │
          │     B/op     │     B/op      vs base               │
Client-10   8.707Ki ± 0%   8.688Ki ± 0%  -0.22% (p=0.000 n=10)

          │  call.txt  │            inline.txt             │
          │ allocs/op  │ allocs/op   vs base               │
Client-10   104.0 ± 0%   103.0 ± 0%  -0.96% (p=0.000 n=10)
```

TLDR: The optimization isn't worth it in context of an entire operation. Consider this comment resolved.

File: internal/mongoutil/mongoutil.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply internal/mongoutil/mongoutil.go:52]: Agreed, benchmarking the version with and without the `reflect.ValueOf` call shows almost no difference, even on Go 1.18. Using `reflect.ValueOf().IsNil` to detect nil interface pointers is a more robust solution than having to do that nil check in every `OptionsSetters` method, so that seems preferable. Consider this comment resolved.

File: internal/integration/unified/collection_operation_execution.go [hunk 20]
```diff
@@ -1181,20 +1191,26 @@ func executeListSearchIndexes(ctx context.Context, operation *operation) (*opera
...
+			lsiOptsCallback := func(args *options.ListSearchIndexesOptions) error {
+				args.AggregateOptions = &options.AggregateOptions{}
+
+				return bson.Unmarshal(val.Document(), args.AggregateOptions)
 			}
-			opts = append(opts, &options.ListSearchIndexesOptions{
-				AggregateOpts: &opt,
-			})
+
+			opts = append(opts, mongoutil.NewBuilderFromOptions[options.ListSearchIndexesOptions](lsiOpts, lsiOptsCallback))
```
Comment (qingyang-hu) [inline internal/integration/unified/collection_operation_execution.go:1202]: Using as an example here, do we have to keep the type parameter for `NewBuilderFromOptions`?
``` go
mongoutil.NewBuilderFromOptions(lsiOpts, lsiOptsCallback)
```
is syntactically correct. Or, do you think it is more semantically clear with the type parameter?

Comment (prestonvasquez) [reply internal/integration/unified/collection_operation_execution.go]: AFAIK Go cannot infer the type here since it's not actually used in the function signature, rather the pointer of the type is.

Comment (qingyang-hu) [reply internal/integration/unified/collection_operation_execution.go]: All type parameters for `NewOptionsLister` in this PR (including "internal/integration/mtest/mongotest.go", "internal/integration/unified/collection_operation_execution.go", "mongo/client_encryption.go", and "mongo/collection.go") are optional as I've tested since `T` has been in the argument list of it.

Comment (prestonvasquez) [reply internal/integration/unified/collection_operation_execution.go]: This incidentally works now because the POC to move mongo.Options to options.SetterLister in [367aabb](https://github.com/mongodb/mongo-go-driver/pull/1570/commits/367aabb8427fe9f688c752178613a0c02fe395eb) changed the type from `mongoutil.Options` to `any`. If we decide to keep that type as any, then we can make this change. Otherwise you will get an error like this:

```
# go.mongodb.org/mongo-driver/internal/integration [go.mongodb.org/mongo-driver/internal/integration.test]
internal/integration/crud_helpers_test.go:905:53: type *"go.mongodb.org/mongo-driver/mongo/options".UpdateOptionsBuilder of opts does not match mongoutil.OptionsBuilder[T] (cannot infer T)
```

File: mongo/options/encryptoptions.go (hunk 14 above)
Comment (matthewdale) [reply mongo/options/encryptoptions.go:37]: Ah, I see where it's used now 👍 . Consider this comment resolved.

**Review by matthewdale — DISMISSED** [review]
Looks good! 👍

File: mongo/options_test.go
Comment (qingyang-hu) [inline mongo/options_test.go:1]: We should remove this file from the PR.

**Review by qingyang-hu — CHANGES_REQUESTED** [review]
Great job! Let's remove "mongo/options_test.go" before merging the PR.

**Review by qingyang-hu — APPROVED** [review]
👍
